xds/src/test/java/io/grpc/xds/ClientXdsClientDataTest.java

```diff
@@ -773,245 +803,197 @@ public void parseCluster_ringHashLbPolicy_invalidRingSizeConfig_tooLargeRingSize
   }
 
   @Test
-  public void parseServerSideListener_invalidTrafficDirection() {
+  public void parseServerSideListener_invalidTrafficDirection() throws ResourceInvalidException {
     Listener listener =
         Listener.newBuilder()
             .setName("listener1")
             .setTrafficDirection(TrafficDirection.OUTBOUND)
             .build();
-    StructOrError<io.grpc.xds.EnvoyServerProtoData.Listener> struct =
-        ClientXdsClient.parseServerSideListener(listener, null);
-    assertThat(struct.getErrorDetail()).isEqualTo("Listener listener1 is not INBOUND");
+    thrown.expect(ResourceInvalidException.class);
+    thrown.expectMessage("Listener listener1 with invalid traffic direction: OUTBOUND");
+    ClientXdsClient.parseServerSideListener(listener, null, true);
   }
 
   @Test
-  public void parseServerSideListener_listenerFiltersPresent() {
+  public void parseServerSideListener_listenerFiltersPresent() throws ResourceInvalidException {
     Listener listener =
         Listener.newBuilder()
             .setName("listener1")
             .setTrafficDirection(TrafficDirection.INBOUND)
             .addListenerFilters(ListenerFilter.newBuilder().build())
             .build();
-    StructOrError<io.grpc.xds.EnvoyServerProtoData.Listener> struct =
-        ClientXdsClient.parseServerSideListener(listener, null);
-    assertThat(struct.getErrorDetail())
-        .isEqualTo("Listener listener1 cannot have listener_filters");
+    thrown.expect(ResourceInvalidException.class);
```

Comment:
What about https://github.com/grpc/grpc-java/issues/7467 i.e. `ExpectedException` being deprecated? Why introduce it when the original test didn't have it?

Reply:
`ExpectedException` is deprecated but `assertThrows` requires Java 8. That's why we still use `ExpectedException` while suppress the warning. The original `ClientXdsClient.parseServerSideListener` returns `StructOrError` instead of throwing exception, throwing `ResourceInvalidException` is the approach we want to move towards.

Reply:
I thought the recommended approach given all the constraints you cited is to use something like this:
```
    try {
      // test code expected to throw exception
      Assert.fail("no exception thrown");
    } catch (UnsupportedOperationException expected) {
      assertThat(expected)
          .hasMessageThat()
          .isEqualTo("Unsupported ... blah...");
    }
```
Do you not like this approach?